core/drivers/dcp/imx_dcp.c

```diff
@@ -0,0 +1,1035 @@
+		/* Channel 1 */
+		.m = MUTEX_INITIALIZER,
+	},
+	{
+		/* Channel 2 */
+		.m = MUTEX_INITIALIZER,
+	},
+	{
+		/* Channel 3 */
+		.m = MUTEX_INITIALIZER,
+	},
+};
+
+/*
+ * Counter used with clock_spinlock
+ * It avoids:
+ *	- useless clock activation (i.e. when the clock was activated by
+ *	  another channel)
+ *	- to disable clock whereas clock is needed by another channel
+ */
+static unsigned int clock_refcount;
+
+static const struct hashalg hash_alg[3] = {
+	{
+		/* SHA1*/
+		.type = DCP_CONTROL1_HASH_SELECT_SHA1,
+		.size = TEE_SHA1_HASH_SIZE,
+	},
+	{
+		/* SHA256 */
+		.type = DCP_CONTROL1_HASH_SELECT_SHA256,
+		.size = TEE_SHA256_HASH_SIZE,
+	},
+	{
+		/* CRC32 */
+		.type = DCP_CONTROL1_HASH_SELECT_CRC32,
+		.size = DCP_CRC32_SIZE,
+	},
+};
+
+/* Compatible DCP string for i.MX6ULL/ULZ/SL/SLL device tree */
+#ifdef CFG_DT
+static const char *const dt_ctrl_match_table[] = {
+	"fsl,imx28-dcp",
+	"fsl,imx6sl-dcp",
+};
+#endif
+
+/* Memory used for context switching */
+static struct align_buf hw_context_buffer;
+
+/*
+ * Enable/disable the DCP clock
+ *
+ * @enable   Enable the clock if true, disable if false
+ */
+static void dcp_clk_enable(bool enable)
+{
+	uint32_t reg = 0;
+	vaddr_t ccm_base = 0;
+
+	ccm_base = core_mmu_get_va(CCM_BASE, MEM_AREA_IO_SEC);
+	reg = io_read32(ccm_base + CCM_CCGR0);
+
+	if (enable)
+		reg |= DCP_CLK_ENABLE_MASK;
+	else
+		reg &= ~DCP_CLK_ENABLE_MASK;
+
+	io_write32(ccm_base + CCM_CCGR0, reg);
+}
+
+/*
+ * Use spinlock to secure clock register wrtiting
+ * Enable clock if it's not done
+ * Disable clock only if no channel is still using it
+ *
+ * @enable   Enable the clock if true, disable if false
+ */
+static void secure_dcp_clk_enable(bool enable)
+{
+	clock_except = cpu_spin_lock_xsave(&clock_spinlock);
+	if (enable) {
+		clock_refcount++;
+		if (clock_refcount == 1)
+			dcp_clk_enable(true);
+	} else {
+		clock_refcount--;
+		if (!clock_refcount)
+			dcp_clk_enable(false);
+	}
+	cpu_spin_unlock_xrestore(&clock_spinlock, clock_except);
+}
+
+/*
+ * Lock a channel using a mutex
+ *
+ * @chan   DCP channel to lock
+ */
+static TEE_Result dcp_lock_known_channel(enum dcp_channel chan)
+{
+	bool lock = false;
+	unsigned int timeout = 0;
+
+	for (timeout = 0; timeout < DCP_MAX_TIMEOUT; timeout++) {
+		lock = mutex_trylock(&lock_channel[chan].m);
+		if (lock)
+			return TEE_SUCCESS;
+
+		dcp_udelay(10);
+	}
+
+	return TEE_ERROR_BUSY;
+}
+
+/*
+ * Unlock a locked channel
+ *
+ * @chan   DCP channel to unlock
+ */
+static void dcp_unlock_channel(enum dcp_channel chan)
+{
+	if (lock_channel[chan].m.state)
+		mutex_unlock(&lock_channel[chan].m);
+}
+
+/*
+ * Try to find an unlocked channel and lock it
+ *
+ * @channel   [out] Channel locked if the function returns TEE_SUCCESS
+ */
+static TEE_Result dcp_lock_any_ready_channel(enum dcp_channel *channel)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	unsigned int i = 0;
+
+	for (i = 0; i < DCP_NB_CHANNELS; i++) {
+		ret = dcp_lock_known_channel((enum dcp_channel)i);
```

Comment:
This seems a bit inefficient, polling for a while on each channel until you succeed to lock one. Why not poll on all a channels at the same time?

Reply:
I removed the timeout in ```dcp_lock_any_ready_channel()```